tests/kernel/pipe: fix uninitialized semaphore

andyross opened the pull request:
The end_sema k_sem was only initialized on one of the several paths
that used it, leading to some crazy clobber-the-run-queue behavior
that was dependent on linkage order (see the linked bug) when end_sema
and the pipe object were made non-static..

Adding a k_sem_init() call fixes the corrupt issue, but really the
right thing is to use the DEFINE macro, so do that instead.  Note that
that the initializer changes the linkage order too (by putting the
semaphore in a separate segment), so... yeah, it's actually impossible
to prove that this patch in isolation resolves the issue seen without
manual validation.

Fixes #4366

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/pipe: fix uninitialized semaphore

The end_sema k_sem was only initialized on one of the several paths
that used it, leading to some crazy clobber-the-run-queue behavior
that was dependent on linkage order (see the linked bug) when end_sema
and the pipe object were made non-static..

Adding a k_sem_init() call fixes the corrupt issue, but really the
right thing is to use the DEFINE macro, so do that instead.  Note that
that the initializer changes the linkage order too (by putting the
semaphore in a separate segment), so... yeah, it's actually impossible
to prove that this patch in isolation resolves the issue seen without
manual validation.

Issue: https://github.com/zephyrproject-rtos/zephyr/issues/4366

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

tyrantlizard77 reviewed (changes requested):
We might not be out of the woods yet.

CI is reporting that test_pipe_block_put is now failing on Riscv32 and xtensa

Commit:
kernel/mem_pool: Correct n_levels computation for small blocks

The new mem pool implementation has a hard minimum block size of 8
bytes, but the macros to statically compute the number of levels
didn't clamp, leading to invalid small allocations being allowed,
which would then corrupt the list pointers of nearby blocks and/or
overflow the buffer entirely and corrupt other memory.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Yeah, the other bug (AFAICT entirely unrelated to the scheduler mess to do with the semaphore) was a initialization bug in mem_pool.  This test uses a minimum block size of 4 bytes, which is smaller than the hard minimum in the allocator of 8.  A few other tests did that too, and got (correct) runtime failures.  But this one turns out to do an immediate free, so it always gets a "valid" buffer.  Except the buffer was allocated out of an invalid level of "4 byte blocks" which can't work.

andrewboie commented on include/kernel.h:
```diff
@@ -3550,6 +3550,11 @@ struct k_mem_pool {
+#define _MPOOL_MINBLK sizeof(sys_dnode_t)
+
+#define _MPOOL_LVLS(max, min)		\
+	__MPOOL_LVLS((max), (min) >= _MPOOL_MINBLK ? (min) : _MPOOL_MINBLK)
```
At first glance I wonder if it might be better to fail an assertion or something, instead of just increasing the block size passed in.
You know the mem pool code far better than me though, so I defer to your opinion.

andrewboie commented:
checkpatch unhappy, rest of tests passing.
i don't know why it wants parens for parameters that just get passed as params to other macros/functions

```
-:10: ERROR:COMPLEX_MACRO: Macros with complex values should be enclosed in parentheses
#10: FILE: include/kernel.h:3535:
+#define __MPOOL_LVLS(maxsz, minsz)		\
 	(_MPOOL_HAVE_LVL(maxsz, minsz, 0) +	\
 	_MPOOL_HAVE_LVL(maxsz, minsz, 1) +	\
 	_MPOOL_HAVE_LVL(maxsz, minsz, 2) +	\

- total: 1 errors, 0 warnings, 37 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

andyross commented:
Grr.... the syntax actually doesn't change.  Somehow this passed checkpatch once before.  Will update.

As far as checking: the API just allows you to specify a "minimum" block size.  It doesn't make promises about exactly how that would be aligned or padded or how many such blocks would be available (you specify the number of maximum-sized blocks instead).  It seems reasonable to allow users to specify whatever they want, they shouldn't have to be concerned with the internal "true minimum" that it gets padded up to.

andyross commented:
Fixed the checkpatch whine.  Turns out checkpatch is smarter than I thought: it's fine with unparenthesized arguments as long as the macro is never **used** with a non-simple expression.  But when I made the argument a ?: expression (even a parenthesized one) it unmasked the warning.
